doc/designs/quic-design/server/quic-polling.md

```diff
@@ -0,0 +1,1637 @@
+poller method does not interact with an internal poller method (i.e., a method
+set on an SSL object). In particular, no two poller methods ever interact
+directly with one another. This avoids the need for recursive state shadowing
+(where one poll method's retained mode API maintains state and also makes calls
+to another poll method's retained mode API).
+
+**Q. How does this design interact with hierarchical polling?**
+
+A. We assume an application uses its own polling arrangements initially and then
+uses calls to an OpenSSL external polling API (such as `SSL_poll` or a poll
+method) to drill down into what is actually ready, as discussed above. There is
+no issue here. An application can also use OpenSSL polling APIs instead of its
+own, if desired; for example it could create a poll group from the default poll
+method and use it to poll only network sockets, some of which may be from QUIC
+SSL object poll descriptors, and then if needed call SSL_poll to narrow things
+down once something becomes ready.
+
+**Q. Should we support immediate and retained mode in the same API or segregate
+these?**
+
+A. They are in the same API, though we let applications use capability bits
+to report support for only one of these if they wish.
+
+**Q. How do we support extensibility of the poller interface?**
+
+A. Using an extensible function table. An application can set a function
+   pointer to NULL if it does not support it. Capability flags are used to
+   advertise what is supported.
+
+**Q. If an application sets a poll method on both an event leader and a poll
+   group, what happens?**
+
+A. Setting a poll method on an event leader provides a mechanism used for internal
+blocking when making blocking calls. It is never used currently if no QUIC SSL
+object in the QUIC domain isn't used in blocking mode (though this isn't a
+contractual guarantee and we might do so in future for fast identification of
+what we need to handle if we handle multiple OS-level sockets in future).
+
+Setting a poll method on a poll group provides a mechanism used for polling
+using that event group. Note that a custom poll method configured on a SSL
+object is **not** used for the translation process performed by a poll group,
+even when polling that SSL object. Translation is driven by
+`SSL_get_[rw]poll_descriptor`.
+
+**Q. What if different poll methods are configured on different event leaders
+     (QUIC domains) and an application then tries to poll them all?**
+
+A. Because the poll method configured on an event leader is ignored in favour of
+the poll method directly invoked, there is no conflict here. The poll method
+handles all polling when it is specifically invoked.
+
+**Q. Where should the responsibility for poll descriptor translation lie?**
+
+A. With the poll method or poll group being called at the time.
+
+**Q. What method does `SSL_poll` use?**
+
+A. It uses the default poll method. If an application wishes to use a different
```

Comment:
Why not the method set by `SSL_set1_poll_method()`? This seems counter-intuitive.

Reply:
What happens if you SSL_poll on multiple different SSL objects with different poll methods set?

Poll methods don't (and can't) really interact with one another because ultimately a poll method must culminate in some single blocking OS API call. It's hard to "compose" the act of waiting in this regard. So any given poll call necessarily involves only one poll method, and if we get asked to wait on some arbitrarily diverse set of SSL objects, it needs to be unambiguous which method is used.